lib/Dialect/AIE/Transforms/AIEAssignLockIDs.cpp

```diff
@@ -40,50 +40,91 @@ struct AIEAssignLockIDsPass : AIEAssignLockIDsBase<AIEAssignLockIDsPass> {
     OpBuilder rewriter = OpBuilder::atBlockEnd(device.getBody());
 
-    std::map<Operation *, std::pair<int, std::set<int>>> tileToLastID;
-
-    // The first pass scans for and stores the existing lockIDs. This data is
-    // stored in a map with the lock’s tile operation as the key, while the
-    // value to the key is a pair with the current potential lockID and a set
-    // that stores the currently assigned lockIDs.
-    for (auto lock : device.getOps<LockOp>()) {
-      if (lock.getLockID().has_value()) {
-        Operation *lock_tile = lock.getTile().getDefiningOp();
-        tileToLastID[lock_tile].first = 0;
-        tileToLastID[lock_tile].second.insert(lock.getLockIDValue());
-      }
-    }
+    // Map from tiles to all their lock ops (stored by ID) which have been
+    // assigned to locks.
+    DenseMap<TileOp, DenseSet<int>> tileToAssignedLocks;
+
+    // Map from tiles to all lock ops which have not been assigned to locks.
+    DenseMap<TileOp, SmallVector<LockOp>> tileToUnassignedLocks;
+
+    // Separate lock ops into assigned and unassigned ops, and store by tile.
+    for (auto lockOp : device.getOps<LockOp>()) {
+
+      auto tileOp = lockOp.getTileOp();
+      bool isAssigned = lockOp.getLockID().has_value();
 
-    // The second pass scans for locks with no lockIDs and assigns locks.
-    for (auto lock : device.getOps<LockOp>()) {
-      Operation *lock_tile = lock.getTile().getDefiningOp();
-      if (!lock.getLockID().has_value()) {
-        if (tileToLastID.find(lock_tile) == tileToLastID.end()) {
-          // If the tile operation corresponding to the lock does not exist in
-          // the data structure, initialize the lockID with 0 with an empty set.
-          tileToLastID[lock_tile].first = 0;
-        } else if (tileToLastID[lock_tile].first < 15) {
-          // If the tile operation of the lock exists, the potential lockID is
-          // checked with the set containing occupied lockIDs until a lockID
-          // that is free is found.
-          int potential_ID = tileToLastID[lock_tile].first;
-          while (true) {
-            if (tileToLastID[lock_tile].second.find(potential_ID) !=
-                tileToLastID[lock_tile].second.end())
-              potential_ID++;
-            else
-              break;
+      // Append to set of assigned locks.
+      if (isAssigned) {
+        auto lockID = lockOp.getLockID().value();
+        auto assignedLocksIter = tileToAssignedLocks.find(tileOp);
+        if (assignedLocksIter == tileToAssignedLocks.end()) {
+          tileToAssignedLocks.insert({tileOp, {lockID}});
+        } else {
+          if (assignedLocksIter->second.find(lockID) !=
+              assignedLocksIter->second.end()) {
+            auto diag = lockOp->emitOpError("is assigned to the same lock (")
+                        << lockID << ") as another op.";
+            diag.attachNote(tileOp.getLoc())
+                << "tile has lock ops assigned to same lock.";
```

Comment:
Could just be "in this tile", I think.  (i.e. a continuation of the previous statement)

Reply:
So, drop the number of available locks? 

Currently it is 
```
../test/assign-lock-ids/test.mlir:24:10: error: 'AIE.lock' op not allocated a lock.
  %l15 = AIE.lock(%tMemTile)
         ^
../test/assign-lock-ids/test.mlir:24:10: note: see current operation: %16 = "AIE.lock"(%0) : (index) -> index
../test/assign-lock-ids/test.mlir:7:15: note: tile has only 16 locks available.
  %tMemTile = AIE.tile(4,4)
  ```
  
  With the change (if I understand your suggestion) it is
  
  ```
  ../test/assign-lock-ids/test.mlir:24:10: error: 'AIE.lock' op not allocated a lock.
  %l15 = AIE.lock(%tMemTile)
         ^
../test/assign-lock-ids/test.mlir:24:10: note: see current operation: %16 = "AIE.lock"(%0) : (index) -> index
../test/assign-lock-ids/test.mlir:7:15: note: in this tile
  %tMemTile = AIE.tile(4,4)
  ```
  
  My default is to add any information which might be useful, and the number of locks on this tile might be?

Reply:
Ew yeah good point.  maybe "because only 16 locks available in this tile"?